src/utils/tr-fetch.ts

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2019 Transposit Corporation. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { APIError } from "../errors/APIError";
+
+/**
+ * trfetch() is a thin-wrapper around native fetch().
+ * It treats non-2XX responses as failures.
+ * It assumes response bodies should be parsed as JSON.
+ */
+export function trfetch<T>(input: RequestInfo, init?: RequestInit): Promise<T> {
+  return fetch(input, init).then((response: Response) => {
+    if (response.ok) {
+      return response.json().then(
+        x => x as T,
+        () => {
+          throw makeInternalError(response);
+        },
```

Comment:
Where did we land on having an OperationError to differentiate "something went wrong processing your request" from "something went wrong in running your operation"?

Reply:
I think we were pro it for actionable operation errors (like you're missing a credential, or your javascript code threw an error server-side).

I'm not clear on whether execution returns error ids that can be used to differentiate between these cases.

Reply:
You can probably disregard this; for some reason I misunderstood this code and thought this was handling that case.